### ISSUE TITLE
Attribute (`__getitem__`) differences for PyROOT objects in ROOT master

### ISSUE THREAD

**Issue by amete** [root-project/root#15161]
### Check duplicate issues.

- [ ] Checked for duplicates

### Description

In `master` it seems like objects on the `PyROOT` side always have `__getitem__` defined. Here is a simple reproducer:

```
$ python -c "import ROOT; ROOT.gInterpreter.LoadText('struct foo{};'); print(hasattr(ROOT.foo(), '__getitem__'));"
```

In version `v6.28.12` this returns `false` whereas in `master` it returns `true`. This seems to be coming from `cppyy`:

```python
import cppyy

cppyy.cppdef("""
  struct foo {};
""")

bar = cppyy.gbl.foo()
print(f"cppyy version {cppyy.__version__}")
print(f"Does foo have __getitem__()? {hasattr(bar, '__getitem__')}")
This gives:
```

This returns:

```
cppyy version 1.6.2
Does foo have __getitem__()? False
```

vs

```
cppyy version 3.1.2
Does foo have __getitem__()? True
```

This change in behavior causes the object, which clearly doesn't support e.g. indexing, to pass the sequence check [(docs)](https://docs.python.org/3/c-api/sequence.html), which creates other issues elsewhere in the ATLAS code.

We briefly discussed this on Mattermost w/ @vepadulano but I wanted to create an issue before I forget 😄 Any ideas?

### Reproducer

Provided in the description

### ROOT version

heads/master@v6-31-01-1576-g7784644928

### Installation method

LCG dev3 nightly build

### Operating system

Linux

**Comment by guitargeek** [comment]
Hi! The new cppyy implements `__getitem__` for every CPPInstance to support subscripting of C-style arrays, like this:
```Python
import cppyy

cppyy.cppdef("""
  struct Foo {
  Foo(int x) : _x{x} {}
  int _x = 0;
  };
""")

v = cppyy.gbl.vector["Foo"]([3, 2, 1])

d = v.data()

print(d[2]._x)
```
I would prefer to not sacrifice this feature, but the `__getitem__` is necessary for it.

Can you replace maybe your `hasattr` check with this?

```python
def is_sequence(obj):
    try:
        obj[0]
        return True
    except TypeError:
        return False
```

This would be backwards compatible, because if `__getitem__` is not there, it will also raise a TypeError.

Let me know if this would work for ATLAS, otherwise let's try to find a different solution.

**Comment by amete** [comment]
Thanks a lot @guitargeek. This is the actual problem is that The Python/C API determines if an object provides sequence control by explicitly checking `__getitem__()` and this behavior is at odds with that:

https://docs.python.org/3.12/c-api/sequence.html#c.PySequence_Check

One could probably use the check in conjunction with the size/length but that's a bit of a kludge.

**Comment by guitargeek** [comment]
I have seen your link, and according to the docs the [PySequence_Check](https://docs.python.org/3.12/c-api/sequence.html#c.PySequence_Check) is a kludge itself:
> Note that it returns 1 for Python classes with a [__getitem__()](https://docs.python.org/3/reference/datamodel.html#object.__getitem__) method, unless they are [dict](https://docs.python.org/3/library/stdtypes.html#dict) subclasses, since in general it is impossible to determine what type of keys the class supports.

There are many other classes in the Python ecosystem that are not dicts and not a sequence, but implement `__getitem__` anyway (Pandas dataframes, uproot files, so many things...).

Where is this used exactly? Internally in CPython and it fails there, or in some ATLAS CPython extension?

It it's in a CPython extension you control, it should not be hard to re-implement the logic with the `except TypeError` with the CPython API. It it's in CPython itself, how does it happen?

In any case, it would be nice to see more info about the usecase (best with some links to code) for me to assess the situation.

**Comment by guitargeek** [comment]
Thanks @wlav!

> I understand where `PySequence_Check` is coming from (there is e.g. a standard Python `iterator` that will try indexing an object from 0 with integers)

Let's ping @amete to ask exactly where it comes from in the context of ATLAS. Knowing how you use PyROOT would help us to make recommendations and if possible improve cppyy for this particular case. I don't think we can converge here without your input.

> Maybe `__getitem__` can be refined to only appear after lookup through `__getattr__`, i.e. to make it fully instance-specific. 

In the original reproducer, the `__getitem__` attribute check is done for the instance, so this wouldn't help there right?

**Comment by amete** [comment]
Thanks @wlav, @guitargeek.

We have a piece of code, [PyROOTInspector](https://gitlab.cern.ch/atlas/athena/-/blob/main/Control/RootUtils/src/pyroot/PyROOTInspector.cxx) that we use to pythonize our data to do some generic operations, such as file comparisons etc. That relies on the C API, hence the issue.

Actually, I recently put in a patch that seems to do the trick, at least from a practical point of view:

https://gitlab.cern.ch/atlas/athena/-/merge_requests/70435/diffs

With this we can recover the same functionality as in the previous `cppyy` versions but I'm in general open to suggestions.

**Comment by wlav** [comment]
> In the original reproducer, the __getitem__ attribute check is done for the instance, so this wouldn't help there right?

Right, I know no way around that. Also, it'd be fine for those cases where a pointer is known to be an array, but like the example above, would fail cases where it's simply *likely* to be an array, as a `__getitem__` would be returned.

> https://gitlab.cern.ch/atlas/athena/-/merge_requests/70435/diffs

That will give you different answers depending on whether the sequence has 0 or N elements, though. But there are plenty more worrisome things in that code. :)

I'm looking at the implementation of `PySequence_Check`, which is:

```
int
PySequence_Check(PyObject *s)
{   
    if (PyDict_Check(s))
        return 0;       
    return Py_TYPE(s)->tp_as_sequence &&
        Py_TYPE(s)->tp_as_sequence->sq_item != NULL;
}
```

The default `__getitem__` is added to the class as part of the `tp_methods` set and I guess it ends up there. A bit strange, though, as `sq_item` is limited to indexing with Py_ssize_t and that's not specified in the method list (only `METH_O`).

But it may give us something to work with as `mp_subscript` is also available. Would have to see whether C++-side `operator[]` properly overrides that and still makes `PySequence_Check` work. It's not directly clear to me that it would, but can give it a try.

**Comment by wlav** [comment]
Hmm, `mp_subscript` mostly works: it fixes the `PySequence_Check` thing, works as expected for other `__getitem__` behaviors. But there's one problem: precisely b/c it no longer passes the sequence check, you can't iterate over a builtin array, represented as a pointer, anymore. I.e., this is no way out.

**Comment by wlav** [comment]
Back to an alternative... Myself, I do the following to find out whether a C++ class had an `operator[]`: https://github.com/wlav/CPyCppyy/blob/master/src/Pythonize.cxx#L38 , but it may not work for you as-is, b/c an important point of that code is to not search for `__getitem__` in base classes, which I think you do want.

Maybe something much simpler would be something along these lines:

```
import cppyy
import cppyy.types

cppyy.cppdef("""\
struct MyStruct1 {};
struct MyStruct2 {
    int operator[](int) { return 42; }
};

#include "CPyCppyy/API.h"

bool CheckSequence(PyObject* obj, PyObject* base) {
   if (CPyCppyy::Instance_Check(obj)) {
       PyObject* gi1 = PyObject_GetAttrString((PyObject*)Py_TYPE(obj), "__getitem__");
       PyObject* gi2 = PyObject_GetAttrString(base, "__getitem__");
       bool ret = gi1 && gi1 != gi2;
       Py_XDECREF(gi2);
       Py_XDECREF(gi1);
       return ret;
   }
   return false;
}
""")

print(cppyy.gbl.CheckSequence(cppyy.gbl.MyStruct1(), cppyy.types.Instance))
print(cppyy.gbl.CheckSequence(cppyy.gbl.MyStruct2(), cppyy.types.Instance))
```

(Where you'd import `cppyy.types.Instance` on the C-side, not pass it as an argument, but this Q&D code was simpler.)

This can be greatly simplified/sped up, by explicitly adding `op_getitem` as a method for `tp_as_sequence` and then do a straight-up pointer comparison. Can probably also expose it as a `CPyCppyy::Sequence_Check` in the API.

**Comment by amete** [comment]
Thanks a lot @wlav! 

> That will give you different answers depending on whether the sequence has 0 or N elements, though.

That's correct. Currently, we'll tag a sequence with 0 elements as not a sequence but the idea in this context is that if there are no elements there is nothing interesting to pythonize so we basically skip that object.

> But there are plenty more worrisome things in that code. :)

:no_mouth: :grinning:

> Can probably also expose it as a `CPyCppyy::Sequence_Check` in the API.

I think this would be the best/ most robust approach. Then we can replace the ad-hoc check we currently do w/ that.

**Comment by wlav** [comment]
Added: https://github.com/wlav/CPyCppyy/commit/a02ceeb4c07a16fe084b8f23e979cb19b9347a05 .

For convenience, it defaults to `PySequence_Check()` for non-cppyy objects.

Note that as the "TODO" in that method says, it's possible to further refine `Sequence_Check` by verifying that the `operator[]()` takes and int-like argument and/or that the class defines a `__len__` of sorts. However, `PySequence_Check` doesn't do any of that either, so I've punted on that for now, but I can add if desired.

The method does not check for `__iter__` as `PySequence_GetItem()` would fail for otherwise iterable containers (such as `std::list`).

Here are some tests showing how I envisage this to work: https://github.com/wlav/cppyy/commit/2ebeacb9aa977497d9dffc8032302f3fc6d418ec
